### PR TITLE
Prometheus adapter tests

### PR DESCRIPTION
**Pull request by paulfantom**
* test availability of v1beta1.metrics.k8s.io api service
* test existence of cpu and memory metrics for nodes
* test existence of cpu and memory metrics for containers inside pods

/assign @s-urbaniak

### CODE REVIEW

File: test/e2e/prometheusadapter_test.go [hunk 1]
```diff
@@ -24,8 +24,88 @@ import (
...
+			lastErr = errors.New("v1beta1.metrics.k8s.io apiservice is not available")
+			return false, nil
+		}
+		return true, nil
+	})
+	if err != nil {
+		if err == wait.ErrWaitTimeout && lastErr != nil {
+			err = lastErr
+		}
+		log.Fatal(err)
```
Comment (brancz) [inline test/e2e/prometheusadapter_test.go]: `t.Fatal`

File: test/e2e/prometheusadapter_test.go [hunk 2]
```diff
@@ -24,8 +24,88 @@ import (
...
+				return false, nil
+			}
+		}
+		return true, nil
+	})
+	if err != nil {
+		if err == wait.ErrWaitTimeout && lastErr != nil {
+			err = lastErr
+		}
+		log.Fatal(err)
```
Comment (brancz) [inline test/e2e/prometheusadapter_test.go]: `t.Fatal`

File: test/e2e/prometheusadapter_test.go [hunk 3]
```diff
@@ -24,8 +24,88 @@ import (
...
+				}
+			}
+		}
+		return true, nil
+	})
+	if err != nil {
+		if err == wait.ErrWaitTimeout && lastErr != nil {
+			err = lastErr
+		}
+		log.Fatal(err)
```
Comment (brancz) [inline test/e2e/prometheusadapter_test.go]: `t.Fatal`

File: test/e2e/prometheusadapter_test.go [hunk 4]
```diff
@@ -24,8 +24,88 @@ import (
...
+	var lastErr error
+	err := wait.Poll(time.Second, 5*time.Minute, func() (bool, error) {
+		nodeMetrics, err := f.MetricsClient.MetricsV1beta1().NodeMetricses().List(metav1.ListOptions{})
+		lastErr = errors.Wrap(err, "getting metrics list failed")
+		if err != nil {
+			return false, nil
+		}
+		for _, item := range nodeMetrics.Items {
+			if item.Usage.Cpu() == nil || item.Usage.Memory() == nil {
+				lastErr = errors.New("node cpu or memory metric not found")
```
Comment (brancz) [inline test/e2e/prometheusadapter_test.go]: this should also verify that we are in fact getting results for all nodes returned by the nodes API

File: test/e2e/prometheusadapter_test.go [hunk 5]
```diff
@@ -24,8 +24,88 @@ import (
...
+	err := wait.Poll(time.Second, 5*time.Minute, func() (bool, error) {
+		podMetrics, err := f.MetricsClient.MetricsV1beta1().PodMetricses(f.Ns).List(metav1.ListOptions{})
+		lastErr = errors.Wrap(err, "getting metrics list failed")
+		if err != nil {
+			return false, nil
+		}
+		for _, pod := range podMetrics.Items {
+			for _, item := range pod.Containers {
+				if item.Usage.Cpu() == nil || item.Usage.Memory() == nil {
+					lastErr = errors.New("container cpu or memory metric not found")
```
Comment (brancz) [inline test/e2e/prometheusadapter_test.go]: this should also verify that we are in fact getting results for all nodes returned by the nodes API

File: test/e2e/prometheusadapter_test.go (hunk 1 above)
Comment (paulfantom) [reply test/e2e/prometheusadapter_test.go]: Everywhere else in this file we are using `log.Fatal` should I also fix those in a follow-up PR?

Comment (brancz) [reply test/e2e/prometheusadapter_test.go]: Whops, yes please! :slightly_smiling_face:

File: test/e2e/prometheusadapter_test.go (hunk 4 above)
Comment (paulfantom) [reply test/e2e/prometheusadapter_test.go]: Added by first checking if both lists are equal length and later iterating over list of nodes.

File: test/e2e/prometheusadapter_test.go (hunk 5 above)
Comment (paulfantom) [reply test/e2e/prometheusadapter_test.go]: Added by first checking if both lists are equal length and later iterating over list of pods.

File: test/e2e/prometheusadapter_test.go [hunk 6]
```diff
@@ -53,43 +53,81 @@ func TestMetricsAPIAvailability(t *testing.T) {
...
+		if len(nodes.Items) != len(nodeMetrics.Items) {
+			lastErr = errors.New("number of nodes doesn't match number of node metrics reported")
+			return false, nil
+		}
 		for _, item := range nodeMetrics.Items {
+			for i, node := range nodes.Items {
+				if node.Name == item.Name {
+					break
+				}
+				if i >= len(nodes.Items) {
```
Comment (brancz) [inline test/e2e/prometheusadapter_test.go]: Not sure this condition is correct. Let's extract "checkIfMetricsExist" to a separate function, then this logic should become a little simpler to read.

Comment (paulfantom) [reply test/e2e/prometheusadapter_test.go]: I have extracted checking if pod/node is in pods/nodes list. This should increase readability a bit.

File: test/e2e/prometheusadapter_test.go [hunk 7]
```diff
@@ -25,10 +25,139 @@ import (
...
+				return false, nil
+			}
+			if item.Usage.Cpu() == nil || item.Usage.Memory() == nil {
+				lastErr = errors.New("node cpu or memory metric not found")
+				return false, nil
+			}
+		}
+		return true, nil
+	})
+	t.Fatal(err)
```
Comment (brancz) [inline test/e2e/prometheusadapter_test.go]: this looks like it shouldn't be here

Comment (paulfantom) [reply test/e2e/prometheusadapter_test.go]: :+1:

File: test/e2e/prometheusadapter_test.go [hunk 8]
```diff
@@ -25,10 +25,138 @@ import (
...
+		if err != nil {
+			return false, nil
+		}
+		if len(pods.Items) != len(podMetrics.Items) {
+			lastErr = errors.New("number of pods doesn't match number of pod metrics reported")
+			return false, nil
+		}
+
+		for _, pod := range podMetrics.Items {
+			if !isPodInPodsList(pod.Name, pods.Items) {
```
Comment (brancz) [inline test/e2e/prometheusadapter_test.go]: a bit of a nit, but this should be the combination of pod and namespace, I realize this is highly unlikely, but let's not leave the possibility open

Comment (paulfantom) [reply test/e2e/prometheusadapter_test.go]: Added with a workaround for all namespaces (when namespace == "").

Comment (brancz) [reply test/e2e/prometheusadapter_test.go]: I don't think that'll work. That should be `pod.Namespace`. Using `""` is just for _selecting_ all metrics, the calls do return pods with their respective namespace set. What I meant with the above comment was we should make sure the pod name + namespace combination is checked for as there can be pods with the same name in different namespaces.

Comment (paulfantom) [reply test/e2e/prometheusadapter_test.go]: Sorry, I misunderstood you. Now it should be ok :)

**Comment by brancz** [comment]
this may need a rebase, I'm not 100% sure

/retest

**Comment by brancz** [comment]
Yeah judging by the CI failure, this needs a rebase.

**Comment by brancz** [comment]
CI failure:

> 	prometheusadapter_test.go:156: number of pods doesn't match number of pod metrics reported

looks like something is off, let's investigate (this may actually be a race with other tests running on the cluster, so we could scope this to the `openshift-monitoring` namespace, as we know the exact pods that should be running there)

**Comment by paulfantom** [comment]
rebased

**Comment by brancz** [comment]
Looking at the test runs, I think this might actually be the pod test itself, that is failing here.

**Comment by paulfantom** [comment]
I will investigate it a bit more.

/hold

**Comment by paulfantom** [comment]
I found what was going on. Seems like a list of pods from core API also has pods from bootstrapping process (42 finished pods) and those are not present when getting a list of pods from metrics API. To fix this I filtered pods by their state (last commit) and in my local tests it seems to work. I am also open to other ideas of solving this.

**Comment by brancz** [comment]
Nice catch!

/lgtm

**Pull request merged**